classes/issue/Collector.inc.php

```diff
@@ -0,0 +1,460 @@
+    {
+        $this->titles = $titles;
+        return $this;
+    }
+
+    /**
+     * Set query search phrase
+     *
+     * @return $this
+     */
+    public function searchPhrase(string $phrase): self
+    {
+        $this->searchPhrase = $phrase;
+        return $this;
+    }
+
+    /**
+     * Limit the number of objects retrieved
+     *
+     * @return $this
+     */
+    public function limit(int $count): self
+    {
+        $this->count = $count;
+        return $this;
+    }
+
+    /**
+     * Offset the number of objects retrieved, for example to
+     * retrieve the second page of contents
+     *
+     * @return $this
+     */
+    public function offset(int $offset): self
+    {
+        $this->offset = $offset;
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getQueryBuilder(): Builder
+    {
+        $this->columns[] = 'i.*';
+        $q = DB::table($this->dao->table, 'i')
+            ->leftJoin('issue_settings as iss', 'i.issue_id', '=', 'iss.issue_id')
+            ->leftJoin('custom_issue_orders as o', 'o.issue_id', '=', 'i.issue_id')
+            // TODO: Is this still necessary (not just for search phrase, but also others)
+            ->groupBy('i.issue_id', $this->orderColumn);
+
+        // Issue titles (exact matches)
+        foreach ($this->titles as $key => $title) {
+            $tableId = 'iss' . $key;
+            $q->join('issue_settings as ' . $tableId, function (JoinClause $join) use ($tableId, $title) {
+                $join->on('i.issue_id', '=', $tableId . '.issue_id')
+                    ->where($tableId . '.setting_name', '=', 'title')
+                    ->where($tableId . '.setting_value', '=', $title);
+            });
+        }
+
+        // Context
+        // Never permit a query without a context_id unless the PKPApplication::CONTEXT_ID_ALL wildcard
+        // has been explicitly set
+        if (is_null($this->contextId)) {
+            $q->where('i.journal_id', '=', PKPApplication::CONTEXT_ID_NONE);
+        } elseif ($this->contextId !== PKPApplication::CONTEXT_ID_ALL) {
+            $q->where('i.journal_id', '=', $this->contextId);
+        }
+
+        // Current
+        // NB: This should not really be used in combination with anything other than
+        // filterByContext
+        if ($this->isCurrent) {
+            $q->where('i.current', '=', 1);
+        }
+
+        // Published
+        if (!is_null($this->isPublished)) {
+            $q->where('i.published', $this->isPublished ? 1 : 0);
+        }
+
+        // Volumes
+        if (!is_null($this->volumes)) {
+            $q->whereIn('i.volume', $this->volumes);
+        }
+
+        // Numbers
+        if (!is_null($this->numbers)) {
+            $q->whereIn('i.number', $this->numbers);
+        }
+
+        // Years
+        if (!is_null($this->years)) {
+            $q->whereIn('i.year', $this->years);
+        }
+
+        // Issue ids
+        if (!empty($this->issueIds)) {
```

Comment:
I'd like to firm up our handling of "falsy" variables as we start to make more use of lists in place of single IDs. I documented this in Slack but I'll put it here for when that disappears:

For conditional SQL, we've historically used PHP `if` statements and built a `$params` array...
```
$params = [];
if ($contextId) $params[] = (int) $contextId;
return new DAOResultFactory($this->retrieve(
    'SELECT ... WHERE ...' .
    ($contextId ? ' AND context_id = ?' : ''),
    $params));
```

Some problems with this:

- It means dealing with $params twice, once when building the $params array and again when building the SQL query. Lots of conditions = messy!
- It's messy about null vs false vs 0 etc.

Laravel query builders improve on this:

```
$query = DB::table('...');
if ($contextId) $query->where('context_id', '=', $contextId);
```

I propose two additional patterns:

- Use Laravel's conditional SQL tool (->when)
- Get strict about "falsy" values. Use null to distinguish between "we don't care" and "we do but no acceptable values are listed". (See below.)

An example using the proposed patterns:

```
$query = DB::table('...')
    ->when($this->contextId !== null, function($query) {
        $query->where('context_id', '=', $this->contextId);
    });
```

This avoids a gotcha, as empty lists are falsy could be misinterpreted as "we don't care". Consider calling code:

```
// We use two conditions to build up a list of allowed user groups:
// 1. Get a list of the current user's groups from the database
$allowedUserGroupsA = [1, 2, 3];
// 2. Get a list of groups the user has specified via the UI (e.g. checkboxes)
$allowedUserGroupsB = [4, 5, 6];
// Combine the two conditions using array_intersect.
$allowedUserGroups = array_intersect($allowedUserGroupsA, $allowedUserGroupsB);
// This will result in an empty list, meaning none of the specified user groups were allowed!
$collector->filterByUserGroupIds($allowedUserGroups);
// We would expect the DAO to return no results.
return $dao->getMany($collector);
```

However, if the collector follows this pattern...

```
// BAD PATTERN BELOW: Empty lists are skipped
$query = DB::table('...')
    ->when($this->userGroupIds, function($query, $userGroupIds) {
        $query->whereIn('user_group_ids', $userGroupIds);
    });
```

...then the `[]` value of `$userGroupIds` means the condition will not be added at all and all user groups will be returned! (Note that this will often mean avoiding the use of the second parameter in lambda functions to `$query->when`. Use `$this` or a `use` clause instead.)

As we start to rely more on arrays when specifying conditions, this kind of gotcha will become more prominent!

So to summarize...

1. In your collector class, initialize constraints to null. Think of this as a reserved value for "we don't care".
   ```
   /** ?array A list of context IDs to allow */
   var $contextIds = null;
   ```
2. When calling the filter function, an array is required, but it could be empty:
   ```
   function filterByContextIds(array $contextIds) {
       $this->contextIds = $contextIds;
   }
   ```
3. When building the query, make sure to treat null values carefully.
   ```
   $query = DB::table(...)
      ->when($this->contextIds !== null, function($query) {
           $query->whereIn('context_id', $this->contextIds
       });
   ```